chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -155,79 +166,72 @@ pub struct Tree {
         time: Timespec,
-        _cs: CipherSuite,
+        cs: CipherSuite,
     ) -> Result<(), kp::Error> {
         let leaf_len =
             LeafSize::from_nodes(NodeSize(nodes.len())).ok_or(kp::Error::TreeIntegrityError)?;
+
+        // leaf should be even position, and parent should be odd position
+        for (i, node) in nodes.iter().enumerate() {
+            if matches!(node, Node::Leaf(_)) && i % 2 != 0 {
+                return Err(kp::Error::TreeIntegrityError);
+            }
+            if matches!(node, Node::Parent(_)) && i % 2 != 1 {
+                return Err(kp::Error::TreeIntegrityError);
+            }
+        }
+
         for (i, node) in nodes.iter().enumerate() {
-            let x = NodeSize(i);
             match node {
-                Some(Node::Leaf(Some(kp))) => {
-                    // leaf should be even position
-                    if i % 2 != 0 {
-                        return Err(kp::Error::TreeIntegrityError);
-                    }
+                Node::Leaf(Some(kp)) => {
                     // "For each non-empty leaf node, verify the signature on the KeyPackage."
                     if let Err(e) = kp.verify(ra_verifier, time) {
                         return Err(e);
                     }
                 }
-                Some(Node::Parent(Some(_pn))) => {
+                Node::Parent(Some(_)) => {
                     // "For each non-empty parent node, verify that exactly one of the node's children are non-empty
                     // and have the hash of this node set as their parent_hash value (if the child is another parent)
                     // or has a parent_hash extension in the KeyPackage containing the same value (if the child is a leaf)."
-                    let left_pos = x.left().unwrap();
-                    let right_pos = x.right(leaf_len).unwrap();
-                    if right_pos.0 >= nodes.len() {
-                        return Err(kp::Error::TreeIntegrityError);
-                    }
-                    // FIXME: compute hash + check
-                    match (&nodes[left_pos.0], &nodes[right_pos.0]) {
-                        (None, Some(_right)) => {}
-                        (Some(_left), None) => {}
+                    let x = NodeSize(i);
+                    let left_pos = x.left().expect("invalid parent");
+                    let right_pos = x.right(leaf_len).expect("invalid parent");
+                    let parent_hash = match (
+                        nodes[left_pos.0].parent_hash(),
+                        nodes[right_pos.0].parent_hash(),
+                    ) {
+                        (None, Some(hash)) => hash,
+                        (Some(hash), None) => hash,
                         _ => {
                             return Err(kp::Error::TreeIntegrityError);
                         }
+                    };
+                    if parent_hash != node_hash(&nodes, cs, x, leaf_len) {
```

Comment:
`.ct_eq` for sanity -- normal `!=` short-circuits / has a timing-leak

Reply:
done